common/sphinx.c

```diff
@@ -669,3 +754,94 @@ u8 *unwrap_onionreply(const tal_t *ctx,
 		return tal_free(final);
 	return final;
 }
+
+struct onionpacket *sphinx_decompress(const tal_t *ctx,
+				      const struct sphinx_compressed_onion *src,
+				      const struct secret *shared_secret)
+{
+	struct onionpacket *res = tal(ctx, struct onionpacket);
+	size_t srclen = tal_bytelen(src->routinginfo);
+	size_t prefill_size = ROUTING_INFO_SIZE - srclen;
+
+	res->version = src->version;
+	res->ephemeralkey = src->ephemeralkey;
+	memcpy(res->mac, src->mac, HMAC_SIZE);
+
+	/* Decompress routinginfo by copying the unmodified prefix, setting
+	 * the compressed suffix to 0x00 bytes and then xoring the obfuscation
+	 * stream in place. */
+	memset(res->routinginfo, 0, ROUTING_INFO_SIZE);
+	memcpy(res->routinginfo, src->routinginfo, srclen);
+	sphinx_prefill_stream_xor(res->routinginfo + srclen, prefill_size,
+				  shared_secret);
+
+	return res;
+}
+
+struct sphinx_compressed_onion *
+sphinx_compress(const tal_t *ctx, const struct onionpacket *packet,
+		const struct sphinx_path *path)
+{
+	struct sphinx_compressed_onion *res;
+	size_t payloads_size = sphinx_path_payloads_size(path);
+
+	/* We can't compress an onion that doesn't have a rendez-vous node. */
+	if (path->rendezvous_id == NULL)
+		return NULL;
+
+	res = tal(ctx, struct sphinx_compressed_onion);
+	res->version = packet->version;
+	res->ephemeralkey = packet->ephemeralkey;
+	memcpy(res->mac, packet->mac, HMAC_SIZE);
+
+	res->routinginfo = tal_arr(res, u8, payloads_size);
+	memcpy(res->routinginfo, packet->routinginfo, payloads_size);
+
+	return res;
+}
+
+u8 *sphinx_compressed_onion_serialize(const tal_t *ctx, const struct sphinx_compressed_onion *onion)
+{
+	size_t routelen = tal_bytelen(onion->routinginfo);
+	size_t len = VERSION_SIZE + PUBKEY_SIZE + routelen + HMAC_SIZE;
+	u8 *dst = tal_arr(ctx, u8, len);
+	u8 der[PUBKEY_CMPR_LEN];
+	int p = 0;
+
+	pubkey_to_der(der, &onion->ephemeralkey);
+
+	write_buffer(dst, &onion->version, VERSION_SIZE, &p);
+	write_buffer(dst, der, PUBKEY_SIZE, &p);
+	write_buffer(dst, onion->routinginfo, routelen, &p);
+	write_buffer(dst, onion->mac, HMAC_SIZE, &p);
+
+	assert(p == len);
+	return dst;
+}
+
+struct sphinx_compressed_onion *
+sphinx_compressed_onion_deserialize(const tal_t *ctx, const u8 *src)
+{
+	const u8 *cursor = src;
+	size_t max = tal_bytelen(src);
+	size_t routelen = max - VERSION_SIZE - PUBKEY_SIZE - HMAC_SIZE;
+	struct sphinx_compressed_onion *dst =
+	    tal(ctx, struct sphinx_compressed_onion);
+
+	/* This is not a compressed onion, so let's not parse it. */
+	if (routelen > ROUTING_INFO_SIZE)
+		return tal_free(dst);
+
+	dst->version = fromwire_u8(&cursor, &max);
+	if (dst->version != 0x00)
+		return tal_free(dst);
+
+	fromwire_pubkey(&cursor, &max, &dst->ephemeralkey);
+
+	dst->routinginfo = tal_arr(dst, u8, routelen);
+	fromwire_u8_array(&cursor, &max, dst->routinginfo, routelen);
+	fromwire_u8_array(&cursor, &max, dst->mac, HMAC_SIZE);
+
+	assert(max == 0);
```

Comment:
Good point, I like `fromwire_tal_bytes`, however it doesn't address the full issue, since that check isn't just there to avoid allocating a huge array, but also to enforce the logic that an onion that is larger than an uncompressed onion can't possibly be a compressed onion.

I added an initial check that the total compressed onion length is smaller than the total length of an uncompressed one, no longer computing `routelen`